### PR TITLE
Fix "Couldn't find branch" error when checking out a PR from a fork

### PR DESCRIPTION
**Pull request by niik**
## Description

When checking out a pull request from a fork prior to #11242 we attempted to look up whether a local branch existed already by looking for a branch named `pr/NUMBER`. This was brittle as it's possible that the user had already checked out the remote branch under another name, or conversely, that the user had a branch named "pr/NUMBER" which did not track the corresponding remote branch (@sergiou87 is fixing on the cause of the latter in https://github.com/desktop/desktop/pull/11266).

Unfortunately my testing of #11242 didn't uncover the fact that the `allBranches` property on `gitStore` didn't actually contain all branches. Forked remote branches were [intentionally excluded](https://github.com/desktop/desktop/blob/49a876e2250b208dc63fd9b7bbf272fffc1491b5/app/src/lib/git/for-each-ref.ts#L100-L106) in order not to clutter the interface. Not cluttering the interface is great but it's applied at too low of a level here.

In this PR I've lifted the filtering of remote fork branches to the presentation layer and audited all the dependents of `getBranches`.

A notable refactor here is the rename of `allBranches` to `branches` in `ICompareState`, that was done because `allBranches` was misleading given that it filtered not only remote fork branches but also the current branch.

I've also had to update the logic of the ahead-behind-updater so that it won't start calculating ahead/behind information for forked remote branches which aren't displayed in the UI (note that this will all be obsolete once #11160 lands).

## Release notes

Notes: [Fixed] Checking out a pull request from a fork for the first time now correctly identifies the remote branch to create the branch from.

### CODE REVIEW

File: app/src/lib/app-state.ts
Comment (sergiou87) [inline app/src/lib/app-state.ts]: > except … the **default** branch

Is it the **default** branch or the **current** branch? This code is using the **current** branch the array: https://github.com/desktop/desktop/pull/11267/files#diff-b238c18a99c08ab800e422715cc02105067f5f1394027f489e52d7fb15d6ca74R1030

🤔

File: app/src/lib/stores/helpers/ahead-behind-updater.ts
Comment (sergiou87) [inline app/src/lib/stores/helpers/ahead-behind-updater.ts]: If I'm following this PR's logic correctly, shouldn't this filter *out* Desktop fork remote branches? 🤔 
```suggestion
      if (b && !b.isDesktopForkRemoteBranch && !cache.has(from, b.tip.sha)) {
```

File: app/src/lib/stores/helpers/ahead-behind-updater.ts
Comment (sergiou87) [inline app/src/lib/stores/helpers/ahead-behind-updater.ts:114]: Assuming I understand the intent of this, and given I needed to look for more implementation details, it's probably worth commenting here that we want recent branches to be scheduled first (since they show up first in the UI), and it's not a problem scheduling them again later with `allBranches` because they'll be in cache already and no extra work will be done.

Does that sound good?

File: app/src/lib/app-state.ts
Comment (niik) [reply app/src/lib/app-state.ts]: Yeah you're absolutely right, should be the **current branch**

File: app/src/lib/stores/helpers/ahead-behind-updater.ts
Comment (niik) [reply app/src/lib/stores/helpers/ahead-behind-updater.ts:114]: > because they'll be in cache already and no extra work will be done.

And also because the `newRefs` is a set consisting of the tip SHA of each branch they won't actually be scheduled to be refreshed. I'm proposing we don't bother too much with this and instead focus on reviewing and merging #11160 which will remove the ahead-behind-updater entirely.

File: app/src/lib/stores/helpers/ahead-behind-updater.ts
Comment (niik) [reply app/src/lib/stores/helpers/ahead-behind-updater.ts]: One day I'll learn how to computer. Great catch @sergiou87!

File: app/src/lib/stores/helpers/ahead-behind-updater.ts
Comment (sergiou87) [reply app/src/lib/stores/helpers/ahead-behind-updater.ts:114]: Yep, I'm reviewing that one as we speak!

**Review by sergiou87 — APPROVED** [review]
Thanks for those changes! LGTM now :shipit: 

Should we wait until #11160 is merged first?

**Comment by niik** [comment]
> Thanks for those changes! LGTM now :shipit:

thank-you-gif

Feels so good to have someone review my code and catch my mistakes again :heart:!

> Should we wait until #11160 is merged first?

Nah, let's get this in. We'll have a few easy-to-resolve conflicts either way. I'll land this and merge development into #11160

**Pull request merged**
